### DISCUSSION TITLE
PieChart Slices Animation

### DISCUSSION THREAD

**Discussion by wayz9** [recharts/recharts#4098] [Q&A] [answered]
Hey everyone,

Is it possible to animate slices of PieChart to grow or expand? So far I managed to replace activeShape with `<Sector>` component and adjust outerRadius to be slightly bigger which increases hovered size but it happens instantly. Is there a way to animate it, apply some transitions?

NextJS 14;
recharts ^2.11
react ^18

I am using NextJS but the Chart component itself is client side and it's being imported via dynamic with ssr false.

Code for reference;

```ts
<PieChart
  margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
  width={270}
  height={270}
>
  <Pie
    startAngle={90}
    animationDuration={550}
    animationEasing="ease-in-out"
    activeIndex={activeIndex}
    activeShape={(props: any) => (
      <Sector {...props} outerRadius={props.outerRadius + 10} />
    )}
    onMouseEnter={(_, index) => setActiveIndex(index)}
    onMouseLeave={() => setActiveIndex(-1)}
    endAngle={-270}
    dataKey="value"
    data={data}
    cx="50%"
    cy="50%"
    innerRadius={65}
    outerRadius={125}
    blendStroke
    label={CustomLabel}
    labelLine={false}
    className="focus:outline-none"
  />
</PieChart>
```

**Comment by ckifer** [comment]
Hi!

Would you be able to get this reproduced in a sandbox? We can go from there - [link](https://codesandbox.io/p/sandbox/pie-chart-demo-forked-qfyldx?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpszce6a00062e6bg75sxjho%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpszce6a00022e6bjs36heh0%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpszce6a00032e6bvgrzozuv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpszce6a00052e6bhrxbsdef%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpszce6a00022e6bjs36heh0%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpszce6a00012e6bbb3wg7zl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clpszce6a00022e6bjs36heh0%2522%252C%2522activeTabId%2522%253A%2522clpszce6a00012e6bbb3wg7zl%2522%257D%252C%2522clpszce6a00052e6bhrxbsdef%2522%253A%257B%2522id%2522%253A%2522clpszce6a00052e6bhrxbsdef%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpszce6a00042e6b4idjtfu5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpszce6a00042e6b4idjtfu5%2522%257D%252C%2522clpszce6a00032e6bvgrzozuv%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpszce6a00032e6bvgrzozuv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by wayz9** [reply]
The link you sent returns "Sandbox not found", it's probably private.

I created something on my own [here](https://codesandbox.io/p/devbox/22j8c4).

**Reply by ckifer** [reply]
Ah sorry, thanks for the link

**Comment by ckifer** [comment, marked as answer]
Try this [![Edit piechart-slices-animation](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/devbox/piechart-slices-animation-forked-zk57tm?workspaceId=7a1754d3-b261-47b1-9339-2aca93020961&embed=1)

I used `react-smooth` for animation (recharts animation library), but you could use any animation library convenient to you that takes a value from one thing to another with animation (react-spring, framer-motion, etc)

**Reply by wayz9** [reply]
That is actually clever, wow. Thank you for your time and help. ✨

**Reply by ckifer** [reply]
No problem!

**Reply by MartinPit** [reply]
Hey, I'm facing the exact same issue. How did you achieve the animation? I cannot see any changes compared to the original sandbox this was forked from, so there isn't any animation for me.

**Reply by wayz9** [reply]
Hey @MartinPit I'll help you out!

https://github.com/user-attachments/assets/5c445bbd-f18a-4e32-b5c4-61df84e8646c

If this is the effect you are after (video above) then this gist will help you. https://gist.github.com/wayz9/08ea5fbb33bb5d9438a1607e4b743345

In the example above you'll see that slices are rendered manually using Cell component. Each cell has some tailwindcss classes that help with the animation. You can find more about tailwind in the [docs](https://tailwindcss.com/docs/transform-origin).

If you need to adjust the animation to be more prominent (increasing the scale size) you would also need to adjust the outerRadius. Right now it's 95% for outerRadius and 105% for hover:scale.

**Reply by MartinPit** [reply]
Thank you for the reply! This is exactly what I wanted! I'm familiar with tailwind, but I didn't think of rendering my own <Cell> components.

**Reply by wayz9** [reply]
> Thank you for the reply! This is exactly what I wanted! I'm familiar with tailwind, but I didn't think of rendering my own components.

Awesome! I forgot to mention most import part: The `Cell` component comes from recharts. 
https://recharts.org/en-US/api/Cell

I'll rewrite my gist today to be a full working example. Hope this clears some things out for you and for anyone else who may find this useful.

**Reply by MartinPit** [reply]
I will have time to properly look at it this evening. Anyway, thanks again for your quick help!

**Reply by wayz9** [reply]
A ready to go example. 

https://gist.github.com/wayz9/db7787db9eccabc4de3a69671c0fd149

You're welcome, always happy to help.
